### ISSUE TITLE
Blofin - fetch_ohlcv  is not working properly, especially "since"

### ISSUE THREAD

**Issue by AltePfeife** [ccxt/ccxt#24677]
### Operating System

Linux Debian12

### Programming Languages

Python

### CCXT Version

4.4.43

### Description

I could swear it worked last month, but maybe Blofin has changed the API.

In the function you can enter any value under “since”, it only ever returns the last 1000 values. 

Does anyone have a solution to the problem?

### Code

```
  def get_future_ohlcv():

    timeframe = '15m'
    symbol = 'XRP/USDT:USDT'
    since = 1728477926799
    limit = 1000

    fohlcv = exchange.fetch_ohlcv(symbol, timeframe, since, limit)
    fohlcv_df = pd.DataFrame(fohlcv, columns=['time', 'open', 'high', 'low', 'close', 'volume'])
    
    berlin_tz = pytz.timezone('Europe/Berlin')

    
    fohlcv_df['current_time'] = fohlcv_df['time'].apply(
        lambda x: datetime.fromtimestamp(x / 1000, tz=berlin_tz).strftime('%Y-%m-%d %H:%M:%S')
    )

    pprint(fohlcv_df)  

```

**Comment by sc0Vu** [comment]
Hi @AltePfeife Could you use `until` value instead? Base on their documentation (https://docs.blofin.com/index.html#get-candlesticks), I think it would work if use `after` (`until` in ccxt),

**Comment by AltePfeife** [comment]
Yes, I was able to create a workaround with until. but it's a bit tricky because you still have to calculate back to “since”. here's my solution:

`def round_to_next_quarter(timestamp_ms):
   
    timestamp_sec = timestamp_ms / 1000.0
    utc_time = datetime.fromtimestamp(timestamp_sec, tz=timezone.utc)    
   
    berlin_tz = pytz.timezone('Europe/Berlin')
    berlin_time = utc_time.astimezone(berlin_tz)    
    
    total_minutes = berlin_time.hour * 60 + berlin_time.minute  
    next_quarter_minutes = (total_minutes + 15 - total_minutes % 15)  
    
    hours = next_quarter_minutes // 60
    minutes = next_quarter_minutes % 60
    rounded_time = berlin_time.replace(hour=hours, minute=minutes, second=0, microsecond=0)
    
    rounded_time_utc = rounded_time.astimezone(timezone.utc)
    
    rounded_timestamp_ms = int(rounded_time_utc.timestamp() * 1000)
    
    return rounded_timestamp_ms

def get_future_bar_data_blofin(symbol, timeframe, since, limit, exchange):

    nfexchange_instance = nfexchange(exchange)

    rounded_time = round_to_next_quarter(since)
    fixed_time = 900000000 + rounded_time

    params = {
        'until': fixed_time
    }

    fohlcv = nfexchange_instance.fetch_ohlcv(symbol, timeframe, None, limit, params)
    fohlcv_df = pd.DataFrame(fohlcv, columns=['time', 'open', 'high', 'low', 'close', 'volume'])
    
    berlin_tz = pytz.timezone('Europe/Berlin')
  
    fohlcv_df['current_time'] = fohlcv_df['time'].apply(
        lambda x: datetime.fromtimestamp(x / 1000, tz=berlin_tz).strftime('%Y-%m-%d %H:%M:%S')
    )

    return fohlcv_df`

**Comment by Midorivi** [comment]
Can you share any updates or workarounds for this issue?

**Comment by carlotestor** [comment]
Blofin's candles endpoint has no "start from" parameter — it only accepts `after` (a *before this time* cursor) plus `limit`, so `since` alone can't work and you always get the most recent page. That's an API limitation, not a CCXT bug.

You don't need to hand-roll the walk-back though — enable pagination and CCXT drives the `after` cursor for you:

```python
ohlcv = exchange.fetch_ohlcv('XRP/USDT:USDT', '15m', since, 1000, {'paginate': True})
```

Or paginate manually with `until` (mapped to `after`), stepping backwards:

```python
until = exchange.milliseconds()
rows = []
while True:
    page = exchange.fetch_ohlcv(symbol, '15m', None, 100, {'until': until})
    if not page:
        break
    rows = page + rows
    until = page[0][0] - 1          # oldest candle in this page, minus 1ms
    if until <= since:
        break
```

Note `limit` is capped at 100 per request on this endpoint, so the built-in `paginate` option is the easier route. @Midorivi — same answer applies.
